### DISCUSSION TITLE
Extension of the Intel Framework

### DISCUSSION THREAD

**Discussion by chrisanag1985** [zeek/zeek#3256] [Ideas]
#  Intro
Extend the Intel Framework in order for the user can define custom scripts with maybe a hook/event, in the case indicator that does not match with the provided IOCs. 

# Little Research
I am developing a script in which, if an indicator is not found in the Intel list, the user can define some `actions`, e.g. sending the hash or file to a sandbox. 

As it is written now by script, I have to re-insert an indicator against the Intel data, something that is not optimal.

After a little research, I found that when an indicator is extracted is sent to the `Intel::seen` exposed function.

```
function Intel::seen(s: Seen)
	{
	if ( find(s) )
		{
		if ( s?$host )
			{
			s$indicator = cat(s$host);
			s$indicator_type = Intel::ADDR;
			}

		if ( ! s?$node )
			{
			s$node = peer_description;
			}

		if ( have_full_data )
			{
			local items = get_items(s);
			event Intel::match(s, items);
			}
		else
			{
			event Intel::match_remote(s);
			}
		}
	}
```
	
The internal `find(s)` function searches in the datastore to find matches. If the match succeeds you have the `Intel::match` event or `Intel::extend_match` hook where the user can add custom code. But if there is no match you cannot add your code. 

Hence I propose adding a point where the user can add code when there is no match and/or create an exposed function `find` which returns for example T/F is it finds the indicator in the datastore.

**Comment by awelzel** [comment]
Thanks @chrisanag1985  - I'm closing this. There is now a `Intel::seen_policy` hook that can be used to hook into `Intel::seen()`.

https://github.com/zeek/zeek/pull/3571
